bokehjs/src/coffee/models/tools/inspectors/hover_tool.coffee

```diff
@@ -175,112 +195,77 @@ export class HoverToolView extends InspectToolView
     x = xscale.invert(vx)
     y = yscale.invert(vy)
 
+    glyph = renderer_view.glyph
+
     for i in indices['0d'].indices
-      data_x = renderer_view.glyph._x[i+1]
-      data_y = renderer_view.glyph._y[i+1]
+      data_x = glyph._x[i+1]
+      data_y = glyph._y[i+1]
       ii = i
 
       switch @model.line_policy
         when "interp" # and renderer.get_interpolation_hit?
-          [data_x, data_y] = renderer_view.glyph.get_interpolation_hit(i, geometry)
+          [data_x, data_y] = glyph.get_interpolation_hit(i, geometry)
           rx = xscale.compute(data_x)
           ry = yscale.compute(data_y)
 
         when "prev"
-          rx = canvas.sx_to_vx(renderer_view.glyph.sx[i])
-          ry = canvas.sy_to_vy(renderer_view.glyph.sy[i])
+          [[rx, ry], ii] = _line_hit(canvas, glyph.sx, glyph.sy, i)
 
         when "next"
-          rx = canvas.sx_to_vx(renderer_view.glyph.sx[i+1])
-          ry = canvas.sy_to_vy(renderer_view.glyph.sy[i+1])
-          ii = i+1
+          [[rx, ry], ii] = _line_hit(canvas, glyph.sx, glyph.sy, i+1)
 
         when "nearest"
-          d1x = renderer_view.glyph.sx[i]
-          d1y = renderer_view.glyph.sy[i]
-          dist1 = hittest.dist_2_pts(d1x, d1y, sx, sy)
-
-          d2x = renderer_view.glyph.sx[i+1]
-          d2y = renderer_view.glyph.sy[i+1]
-          dist2 = hittest.dist_2_pts(d2x, d2y, sx, sy)
-
-          if dist1 < dist2
-            [sdatax, sdatay] = [d1x, d1y]
-          else
-            [sdatax, sdatay] = [d2x, d2y]
-            ii = i+1
-
-          data_x = renderer_view.glyph._x[i]
-          data_y = renderer_view.glyph._y[i]
-          rx = canvas.sx_to_vx(sdatax)
-          ry = canvas.sy_to_vy(sdatay)
+          [[rx, ry], ii] = _nearest_line_hit(canvas, i, geometry, sx, sy, glyph.sx, glyph.sy)
+          data_x = glyph._x[i]
+          data_y = glyph._y[i]
```

Comment:
These should use the returned `ii` for indexing, so that the data x and data y are correct.

Reply:
Thanks for the catch @clairetang6 I will make the change and merge

Reply:
Should this similarly be `jj`? 
```
          when "nearest"
              [[rx, ry], jj] = _nearest_line_hit(canvas, j, geometry, sx, sy, glyph.sxs[i], glyph.sys[i])
              data_x = glyph._xs[i][j]
              data_y = glyph._ys[i][j]
```

Reply:
Yeah, that should also be `jj`. 

Also, from testing with multiline, I found that there is an extra loop in the code for the 2d selections. The outer loop is through the 1d indices, then for multilines, it loops again through the 2d indices, but shouldn't. The line `for i, [j] of indices['2d'].indices` could be changed to `j = indices['2d'].indices[i]` with the `i` being defined by the outer loop already. Or the loop for the multiline could be pulled out of the loop for the 1d indices.